src/data/list/palindrome.lean

```diff
@@ -0,0 +1,67 @@
+/-
+Copyright (c) 2020 Google LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Wong
+-/
+
+import data.list.basic
+open list
+
+/-!
+# Palindromes
+
+This module defines *palindromes*, lists which are equal to their reverse.
+
+The main result is the `palindrome` inductive type, and its associated `palindrome.rec_on` induction
+principle. Also provided are conversions to and from other equivalent definitions.
+
+## References
+
+* [Pierre Castéran, *On palindromes*][casteran]
+
+[casteran]: https://www.labri.fr/perso/casteran/CoqArt/inductive-prop-chap/palindrome.html
+
+## Tags
+
+palindrome, reverse, induction
+-/
+
+variables {α : Type*}
+
+/--
+`palindrome l` asserts that `l` is a palindrome. This is defined inductively:
+
+* The empty list is a palindrome;
+* A list with one element is a palindrome;
+* Adding the same element to both ends of a palindrome results in a bigger palindrome.
+-/
+inductive palindrome : list α → Prop
+| nil : palindrome []
+| singleton : ∀ x, palindrome [x]
+| cons_concat : ∀ x {l}, palindrome l → palindrome (x :: (l ++ [x]))
+
+namespace palindrome
+
+lemma reverse_eq {l : list α} (p : palindrome l) : reverse l = l :=
+palindrome.rec_on p rfl (λ _, rfl) (λ x l p h, by simp [h])
+
+lemma of_reverse_eq {l : list α} : reverse l = l → palindrome l :=
+begin
+  refine bidirectional_rec_on l (λ _, palindrome.nil) (λ a _, palindrome.singleton a) _,
+  intros x l y hp hr,
+  rw [reverse_cons, reverse_append] at hr,
+  rw head_eq_of_cons_eq hr,
+  have : palindrome l, from hp (append_inj_left' (tail_eq_of_cons_eq hr) rfl),
+  exact palindrome.cons_concat x this
+end
+
+lemma iff_reverse_eq {l : list α} : palindrome l ↔ reverse l = l :=
+iff.intro reverse_eq of_reverse_eq
+
+lemma append_reverse (l : list α) : palindrome (l ++ reverse l) :=
+by { apply of_reverse_eq, rw [reverse_append, reverse_reverse] }
+
+instance (l : list α) : decidable (palindrome l) :=
+decidable_of_iff _ iff_reverse_eq.symm
```

Comment:
Playing devil's advocate here - what's the advantage of this approach vs the more obvious:
```suggestion
def palindrome : (l : list ) := reverse l = l

namespace palindrome

lemma reverse_eq {l : list α} (p : palindrome l) : reverse l = l := rfl

lemma of_reverse_eq {l : list α} : reverse l = l → palindrome l := rfl

lemma iff_reverse_eq {l : list α} : palindrome l ↔ reverse l = l := rfl  -- or similar

lemma append_reverse (l : list α) : palindrome (l ++ reverse l) :=
by rw [reverse_append, reverse_reverse]

instance (l : list α) : decidable (palindrome l) := rfl
```

Would it be simpler to just define a `rec_on` lemma rather than define a new inductive type and show its equivalence to the non-inductive statement>

Reply:
Depending on what you want to do with it, the inductive characterization can be easier to work with, e.g. your version still needs to prove `cons_concat`. Right before your post I edited the commit message to note that this definition might be preferable, and future work will show if it should be changed -- with no dependencies it's easy to switch and hard to tell which is better.